bash scripts: use /usr/bin/env for bash shebangs

grahamc opened the pull request:
### Motivation and Context

Not all systems / distros have a `/bin/bash`, and these scripts are
more difficult to run at development time.

### Description

For example, my system is NixOS which doesn't have a /bin/bash. This
is not a problem for NixOS building ZFS as a package: the build
environment automatically replaces these shebangs with corrected
paths.

The problem is much more annoying at development time: either the
scripts don't run, or I correct them for my local machine and deal with
a perpetually dirty work tree.

There are a handful of other shebangs in this repository which don't
work on my system. This patch is useful on its own specifically for
`commitcheck.sh`, otherwise I can't validate my commits before
submission.

Here are the remaining shebangs which NixOS systems won't have:

       1274 #!/bin/ksh -p
         91 #!/bin/ksh
         89 #! /bin/ksh -p
          2 #!/bin/sed -f
          1 #!/usr/bin/perl -w
          1 #!/usr/bin/ksh
          1 #!/bin/nawk -f

plus this which will create an invalid shebang in
`tests/zfs-tests/tests/functional/mv_files/mv_files_common.kshlib`:

        echo "#!/bin/ksh" > $TEST_BASE_DIR/exitsZero.ksh

I chose to leave those alone for now, and gauge the interest in this
much smaller patch first.

The fixes for these are easy enough by simply using `/usr/bin/env ksh`:

         91 #!/bin/ksh
          1 #!/usr/bin/ksh

The fix for the other set is much trickier. Quoting the GNU coreutils
manual:

    Most operating systems (e.g. GNU/Linux, BSDs) treat all text after
    the first space as a single argument. When using env in a script it
    is thus not possible to specify multiple arguments.

and not all `env`'s support arguments.

Mine (GNU Coreutils 8.31) does, though this feature is new since
April 2018, GNU Coreutils 8.30:
https://git.savannah.gnu.org/cgit/coreutils.git/commit/?id=668306ed86c8c79b0af0db8b9c882654ebb66db2

and worse, requires the -S argument:

    -S, --split-string=S  process and split S into separate arguments;
                          used to pass multiple arguments on shebang
                          lines

Example:

    $ seq 1 2 | $(nix-build '<nixpkgs>' -A coreutils)/bin/env "sort -nr"
    /nix/[...]-coreutils-8.31/bin/env: ‘sort -nr’: No such file or directory
    /nix/[...]-coreutils-8.31/bin/env: use -[v]S to pass options in shebang lines

    $ seq 1 2 | $(nix-build '<nixpkgs>' -A coreutils)/bin/env "-S sort -nr"
    2
    1

GNU Coreutils says FreeBSD's `env` does, though I wonder if FreeBSD's
would be unhappy with the `-S`:
https://www.gnu.org/software/coreutils/manual/html_node/env-invocation.html#env-invocation

BusyBox v1.30.1 does not, and does not have a `-S`-like option:

    $ seq 1 2 | $(nix-build '<nixpkgs>' -A busybox)/bin/env "sort -nr"
    env: can't execute 'sort -nr': No such file or directory

Toybox 0.8.1 also does not, and also does not have a `-S` option:

    $ seq 1 2 | $(nix-build '<nixpkgs>' -A toybox)/bin/env "sort -nr"
    env: exec sort -nr: No such file or directory

---

At any rate, if this patch merges and the remaining ~1,500 are updated,
the much larger patch should probably include a checkstyle-like test
asserting all new shebangs use `/usr/bin/env`. I also don't mind
dealing with NixOS weirdness if the project would prefer that.

Signed-off-by: Graham Christensen <graham@grahamc.com>
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->



### How Has This Been Tested?

Before committing this patch I confirmed there are existing scripts
which use `/usr/bin/env` to locate bash, so I am thinking this is a
safe transformation.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

grahamc commented:
Should I break the URLs to appease the stylechecker?

freqlabs commented:
You'll need to add bash to the list of Linux commands in tests/zfs-tests/include/commands.cfg

behlendorf commented:
> Not all systems / distros have a /bin/bash, and these scripts are more difficult to run at development time.

As I understand it the issue you're want to address here is the fully specified path, not that `bash` dependency itself.  That sounds good, the packaging process should do the appropriate shebang mangling.  But as @freqlabs mentioned you will need to include the following additional change to fix the failed tests.

```diff
diff --git a/tests/zfs-tests/include/commands.cfg b/tests/zfs-tests/include/commands.cfg
index 9d1fd89..abb9baf 100644
--- a/tests/zfs-tests/include/commands.cfg
+++ b/tests/zfs-tests/include/commands.cfg
@@ -132,6 +132,7 @@ export SYSTEM_FILES_FREEBSD='chflags
     uncompress'
 
 export SYSTEM_FILES_LINUX='attr
+    bash
     blkid
     blockdev
     chattr
```

Additionally, you may have noticed we've been systematically converting the existing bash scripts to the more portable bourne shell.  The intent is to eventually remove the bash dependency to be compatible with distributions which only include sh / ash / dash / etc in their minimal installs.  One exception here could be the dracut scripts, since it also depends on bash.  

> 1274 #!/bin/ksh -p

I don't believe we need the `-p` here any longer.  It could probably be dropped without causing any issues to make thing easier.  It's something which was inherited from the original test scripts.

We'd need to decide how to handle the other few remaining shebangs on a case by case basis.

> Should I break the URLs to appease the stylechecker?

No need.  Those are alright, the style checker really should just be a little smarter.

> At any rate, if this patch merges and the remaining ~1,500 are updated, the much larger patch should probably include a checkstyle-like test asserting all new shebangs use /usr/bin/env.

Absolutely, once updated we'll want to update the scripts to enforce this.

grahamc commented:
> > Not all systems / distros have a /bin/bash, and these scripts are more difficult to run at development time.
> 
> As I understand it the issue you're want to address here is the fully specified path, not that `bash` dependency itself.

Exactly.

>  as @freqlabs mentioned you will need to include the following additional change to fix the failed tests.

Thank you for making it so clear where in that file `bash` needed to be added - I've updated my patch to include this change. I didn't add a `Co-authored-by` section since it is pretty trivial, but I would be happy to if that is appropriate.

With regards to switching to the borne shell and dropping `-p` from `ksh`: I haven't yet gone through the process of setting up a way for me to run the ZFS test suite, and I'd definitely need that to feel comfortable making such a significant change.

I've also rebased on to `master` thinking that might fix a build problem. I also checked the failing style build, these are the lines over 72 characters:

```
https://git.savannah.gnu.org/cgit/coreutils.git/commit/?id=668306ed86c8c79b0af0db8b9c882654ebb66db2
    $ seq 1 2 | $(nix-build '<nixpkgs>' -A coreutils)/bin/env "sort -nr"
    /nix/[...]-coreutils-8.31/bin/env: ‘sort -nr’: No such file or directory
    /nix/[...]-coreutils-8.31/bin/env: use -[v]S to pass options in shebang lines
    $ seq 1 2 | $(nix-build '<nixpkgs>' -A coreutils)/bin/env "-S sort -nr"
https://www.gnu.org/software/coreutils/manual/html_node/env-invocation.html#env-invocation
```

I tried to make the non-URL lines as short as possible, but had a hard time shortening them further. If this is important, I could reformat them to fit -- but it becomes a bit awkward I think.

grahamc commented:
I see a couple test builds fail. I'm not sure what it is, but I find it really hard to know if I've caused these failures or not. They _seem_ unrelated, but :sweat_smile: .

freqlabs commented:
Style you can fix:
`error: commit message body contains line over 72 characters`
The other two don't seem to be your fault.

rlaager commented:
The perl -w can be replaced with `/usr/bin/env perl` and then adding `use warnings;` to the top of the perl code. This is apparently preferred anyway, as the man page for perl says:
> The "use warnings" pragma produces some lovely diagnostics. One can
> also use the -w flag, but its use is normally discouraged, because it
> gets applied to all executed Perl code, including that not under your
> control.

If you can't rely on an `env` supporting multiple arguments, you will probably have to fall back to substituting the shebang (i.e. detect the path in ./configure, rename the source files from x to x.in, and build them with substitutions using, e.g. sed).

grahamc commented:
@behlendorf already said to ignore the style failures in this case. I'm glad the other two are unrelated.
